### ISSUE TITLE
javassist.bytecode.DuplicateMemberException: duplicate method: realmGet$id in

### ISSUE THREAD

**Issue by stoleruedgar** [realm/realm-java#3469]
#### Goal

We have 2 Android Library projects. 

'sdk-core' - which has the core functionality which has the core RealmObjects.
'sdk' - extends the 'sdk-core'.

We are trying to create a fat aar for the 'sdk' using fat-aar.gradle from this github project (https://github.com/adwiv/android-fat-aar)

The :sdk-core:transformClassesWithRealmTransformerForRelease passes. It seems that after the classes are bundled together the call to :sdk:transformClassesWithRealmTransformerForRelease fails on a RealmObject that is in the 'sdk-core' -com.proxiwave.core.persistence.model.RealmTarget

#### Expected Results

BUILD Successful

#### Actual Results

```java
Incremental java compilation is an incubating feature.
:clean UP-TO-DATE
:sdk:clean
:sdk-core:clean
:sdk:preBuild UP-TO-DATE
:sdk:preReleaseBuild UP-TO-DATE
:sdk:compileReleaseNdk UP-TO-DATE
:sdk:compileLint
:sdk:copyReleaseLint UP-TO-DATE
:sdk:mergeReleaseShaders
:sdk:compileReleaseShaders
:sdk:embedAssets
:sdk:generateReleaseAssets
:sdk:checkReleaseManifest
:sdk:preDebugAndroidTestBuild UP-TO-DATE
:sdk:preDebugBuild UP-TO-DATE
:sdk:preDebugUnitTestBuild UP-TO-DATE
:sdk:preReleaseUnitTestBuild UP-TO-DATE
:sdk-core:preBuild UP-TO-DATE
:sdk-core:preReleaseBuild UP-TO-DATE
:sdk-core:compileReleaseNdk UP-TO-DATE
:sdk-core:compileLint
:sdk-core:copyReleaseLint UP-TO-DATE
:sdk-core:checkReleaseManifest
:sdk-core:preDebugAndroidTestBuild UP-TO-DATE
:sdk-core:preDebugBuild UP-TO-DATE
:sdk-core:preDebugUnitTestBuild UP-TO-DATE
:sdk-core:preReleaseUnitTestBuild UP-TO-DATE
:sdk-core:prepareComAndroidSupportAnimatedVectorDrawable2340Library
:sdk-core:prepareComAndroidSupportAppcompatV72340Library
:sdk-core:prepareComAndroidSupportSupportV42340Library
:sdk-core:prepareComAndroidSupportSupportVectorDrawable2340Library
:sdk-core:prepareComGetkeepsafeRelinkerRelinker121Library
:sdk-core:prepareIoRealmRealmAndroidLibrary110Library
:sdk-core:prepareOrgAltbeaconAndroidBeaconLibrary281Library
:sdk-core:prepareReleaseDependencies
:sdk-core:compileReleaseAidl
:sdk-core:compileReleaseRenderscript
:sdk-core:generateReleaseBuildConfig
:sdk-core:mergeReleaseShaders
:sdk-core:compileReleaseShaders
:sdk-core:generateReleaseAssets
:sdk-core:mergeReleaseAssets
:sdk-core:generateReleaseResValues UP-TO-DATE
:sdk-core:generateReleaseResources
:sdk-core:mergeReleaseResources
:sdk-core:processReleaseManifest
:sdk-core:processReleaseResources
:sdk-core:generateReleaseSources
:sdk-core:incrementalReleaseJavaCompilationSafeguard
:sdk-core:compileReleaseJavaWithJavac
Note: Version 1.2.0 of Realm is now available: http://static.realm.io/downloads/java/latest
Note: Processing class RealmAd
Note: Processing class RealmBeacon
Note: Processing class RealmCampaign
Note: Processing class RealmCompany
Note: Processing class RealmFloor
Note: Processing class RealmStore
Note: Processing class RealmTarget
Note: Processing class RealmTrigger
Note: Processing module RealmModule
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:sdk-core:extractReleaseAnnotations
:sdk-core:mergeReleaseProguardFiles UP-TO-DATE
:sdk-core:packageReleaseRenderscript UP-TO-DATE
:sdk-core:packageReleaseResources
:sdk-core:transformClassesWithRealmTransformerForRelease
:sdk-core:processReleaseJavaRes UP-TO-DATE
:sdk-core:transformResourcesWithMergeJavaResForRelease
:sdk-core:transformClassesAndResourcesWithSyncLibJarsForRelease
:sdk-core:mergeReleaseJniLibFolders
:sdk-core:transformNative_libsWithMergeJniLibsForRelease
:sdk-core:transformNative_libsWithSyncJniLibsForRelease
:sdk-core:bundleRelease
:sdk:prepareComAndroidSupportAnimatedVectorDrawable2340Library
:sdk:prepareComAndroidSupportAppcompatV72340Library
:sdk:prepareComAndroidSupportSupportV42340Library
:sdk:prepareComAndroidSupportSupportVectorDrawable2340Library
:sdk:prepareComGetkeepsafeRelinkerRelinker121Library
:sdk:prepareComGoogleAndroidGmsPlayServicesBasement920Library
:sdk:prepareComGoogleAndroidGmsPlayServicesTasks920Library
:sdk:prepareComGoogleFirebaseFirebaseCommon920Library
:sdk:prepareComGoogleFirebaseFirebaseIid920Library
:sdk:prepareComGoogleFirebaseFirebaseMessaging920Library
:sdk:prepareIoRealmRealmAndroidLibrary110Library
:sdk:prepareOrgAltbeaconAndroidBeaconLibrary281Library
:sdk:prepareProxiwaveSdkAndroidSdkCoreUnspecifiedLibrary
:sdk:prepareReleaseDependencies
:sdk:mergeReleaseAssets
:sdk:compileReleaseRenderscript
:sdk:generateReleaseResValues UP-TO-DATE
:sdk:generateReleaseResources
:sdk:mergeReleaseResources
:sdk:processReleaseManifest
:sdk:processReleaseResources
:sdk:embedRClassFiles
:sdk:compileReleaseAidl
:sdk:generateReleaseBuildConfig
:sdk:generateReleaseSources
:sdk:incrementalReleaseJavaCompilationSafeguard
:sdk:compileReleaseJavaWithJavac
Note: Version 1.2.0 of Realm is now available: http://static.realm.io/downloads/java/latest
Note: Processing class RealmAddress
Note: Processing class RealmDevice
Note: Processing class RealmEvent
Note: Processing class RealmIdentity
Note: Processing class RealmPhoneNumber
Note: Processing class RealmProfile
Note: Processing class RealmSetting
Note: Processing class RealmWeight
Note: Processing module ProxiWaveModule
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:sdk:embedJavaJars
:sdk:mergeReleaseJniLibFolders
:sdk:embedJniLibs
====== Embedding jni libs in project ':sdk'
======= Copying JNI from build/intermediates/exploded-aar/proxiwave-sdk-android/sdk-core/unspecified/jni
:sdk:embedManifests
========== INFO : Loading library manifest /Users/Edgar/Workspaces/xcode/proxiwave-white-label-android/proxiwave-sdk-android/sdk/build/intermediates/exploded-aar/proxiwave-sdk-android/sdk-core/unspecified/AndroidManifest.xml
========== INFO : Merging main manifest /Users/Edgar/Workspaces/xcode/proxiwave-white-label-android/proxiwave-sdk-android/sdk/build/intermediates/bundles/release/AndroidManifest.orig.xml

========== INFO : Merging library manifest /Users/Edgar/Workspaces/xcode/proxiwave-white-label-android/proxiwave-sdk-android/sdk/build/intermediates/exploded-aar/proxiwave-sdk-android/sdk-core/unspecified/AndroidManifest.xml
========== INFO : Merging manifest with lower AndroidManifest.xml:2:1-26:12
========== INFO : Merging uses-sdk with lower AndroidManifest.xml:7:5-9:41
========== INFO : Merging application with lower AndroidManifest.xml:19:5-24:19
========== INFO : Merging result:SUCCESS
========== INFO : Merged manifest saved to /Users/Edgar/Workspaces/xcode/proxiwave-white-label-android/proxiwave-sdk-android/sdk/build/intermediates/bundles/release/AndroidManifest.xml
:sdk:extractReleaseAnnotations
:sdk:mergeReleaseProguardFiles UP-TO-DATE
:sdk:packageReleaseRenderscript UP-TO-DATE
:sdk:embedProguard
====== Embedding proguard.txt files in project ':sdk'
:sdk:embedLibraryResources
:sdk:packageReleaseResources
:sdk:transformClassesWithRealmTransformerForRelease FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':sdk:transformClassesWithRealmTransformerForRelease'.
> javassist.bytecode.DuplicateMemberException: duplicate method: realmGet$id in com.proxiwave.core.persistence.model.RealmTarget

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':sdk:transformClassesWithRealmTransformerForRelease'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:69)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:66)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:203)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:185)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:66)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:50)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:25)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:110)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
        at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
        at org.gradle.initialization.DefaultGradleLauncher$4.run(DefaultGradleLauncher.java:153)
        at org.gradle.internal.Factories$1.create(Factories.java:22)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:53)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:150)
        at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:32)
        at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:98)
        at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:92)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:63)
        at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:92)
        at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:83)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:99)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:48)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:30)
        at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:81)
        at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:46)
        at org.gradle.launcher.exec.DaemonUsageSuggestingBuildActionExecuter.execute(DaemonUsageSuggestingBuildActionExecuter.java:51)
        at org.gradle.launcher.exec.DaemonUsageSuggestingBuildActionExecuter.execute(DaemonUsageSuggestingBuildActionExecuter.java:28)
        at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:43)
        at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:173)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:239)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:212)
        at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:35)
        at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:24)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:205)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:169)
        at org.gradle.launcher.Main.doAction(Main.java:33)
        at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:55)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:36)
        at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
        at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:30)
        at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:129)
        at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
Caused by: java.lang.RuntimeException: javassist.bytecode.DuplicateMemberException: duplicate method: realmGet$id in com.proxiwave.core.persistence.model.RealmTarget
        at com.android.builder.profile.Recorder$Block.handleException(Recorder.java:54)
        at com.android.builder.profile.ThreadRecorder$1.record(ThreadRecorder.java:57)
        at com.android.builder.profile.ThreadRecorder$1.record(ThreadRecorder.java:47)
        at com.android.build.gradle.internal.pipeline.TransformTask.transform(TransformTask.java:173)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$IncrementalTaskAction.doExecute(AnnotationProcessingTaskFactory.java:245)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:221)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$IncrementalTaskAction.execute(AnnotationProcessingTaskFactory.java:232)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:210)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
        ... 60 more
Caused by: javassist.bytecode.DuplicateMemberException: duplicate method: realmGet$id in com.proxiwave.core.persistence.model.RealmTarget
        at javassist.bytecode.ClassFile.testExistingMethod(ClassFile.java:705)
        at javassist.bytecode.ClassFile.addMethod(ClassFile.java:680)
        at javassist.CtClassType.addMethod(CtClassType.java:1383)
        at io.realm.transformer.BytecodeModifier$_addRealmAccessors_closure1.doCall(BytecodeModifier.groovy:45)
        at io.realm.transformer.BytecodeModifier.addRealmAccessors(BytecodeModifier.groovy:42)
        at io.realm.transformer.BytecodeModifier$addRealmAccessors$0.call(Unknown Source)
        at io.realm.transformer.RealmTransformer$_transform_closure10.doCall(RealmTransformer.groovy:132)
        at io.realm.transformer.RealmTransformer.transform(RealmTransformer.groovy:131)
        at com.android.build.api.transform.Transform.transform(Transform.java:308)
        at com.android.build.gradle.internal.pipeline.TransformTask$3.call(TransformTask.java:178)
        at com.android.build.gradle.internal.pipeline.TransformTask$3.call(TransformTask.java:174)
        at com.android.builder.profile.ThreadRecorder$1.record(ThreadRecorder.java:55)
        ... 69 more


BUILD FAILED

Total time: 18.878 secs
```

#### Steps & Code to Reproduce

./gradlew clean assembleRelease

#### Version of Realm and tooling

Realm version(s): 1.1.0

Android Studio version: 2.2

**Comment by Zhuinden** [comment]
Did you share the RealmObjects via a RealmModule?

**Comment by stoleruedgar** [comment]
@Zhuinden 

Yes.

The 'sdk-core' module has 

```java

@io.realm.annotations.RealmModule(library = true, allClasses = true)
public class RealmModule {
}

```

and the `sdk` module has 

```java

@RealmModule(library = true, allClasses = true)
public class ProxiWaveModule {
}

```

**Comment by Zhuinden** [comment]
aaand you don't have the exact same class in both modules?

**Comment by stoleruedgar** [comment]
@Zhuinden 

The classes.jar from the 'sdk-core' module gets pulled into the 'sdk' module. Seems that realm-plugin has a problem in skipping the classes with realmGet$ and realmSet$ methods.

**Comment by stoleruedgar** [comment]
May I did not explain clearly: we have a gradle script which builds the 'sdk-core' library project and then pulls the classes.jar and merges in all the resources and manifest file into the master 'sdk' library project in order to create a fat aar file. The realm objects from 'sdk-core' get transformed by the realm-android plugin. Because they classes.jar gets pulled into the 'sdk' library project on a second pass of the realm-android plugin the files already transformed are not skipped and thus the

```java 
javassist.bytecode.DuplicateMemberException: duplicate method: realmGet$id in com.proxiwave.core.persistence.model.RealmTarget
```

**Comment by zaki50** [comment]
Transformer seems to check the existing getters/setters to void duplicate method to be added.

https://github.com/realm/realm-java/blob/master/realm-transformer/src/main/groovy/io/realm/transformer/BytecodeModifier.groovy#L44

I'll check if this code is working as expected.

**Comment by zaki50** [comment]
Duplicate checking was not woking as expected.
I made a PR (#3477) to fix this issue.

**Comment by zaki50** [comment]
fixed in 2.0.0
